File: 2807-insert-greatest-common-divisors-in-linked-list/2807-insert-greatest-common-divisors-in-linked-list.py
class Solution:
    def insertGreatestCommonDivisors(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head == None or head.next == None: return head
        
        curr = head
        while curr.next:
            nextNode = curr.next
            gcdVal = self.gcd(curr.val, nextNode.val)
            gcdNode = ListNode(gcdVal, nextNode)
            curr.next = gcdNode
            curr = nextNode
        return head

    def gcd(self, a: int, b: int) -> int:
        # ensure that b is always smaller
        if b > a: return gcd(b, a)

        while b:
            temp = b
            b = a % b
            a = temp
        return a

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

File: 2807-insert-greatest-common-divisors-in-linked-list/2807-insert-greatest-common-divisors-in-linked-list.cs
public class Solution {
    public ListNode InsertGreatestCommonDivisors(ListNode head) {
        if(head == null || head.next == null) return head;
        
        var curr = head;
        while(curr.next != null){
            var next = curr.next;
            var gcdVal = GCD(curr.val, next.val);
            var gcdNode = new ListNode(gcdVal, next);
            curr.next = gcdNode;
            curr = next;
        }

        return head;
    }

    private int GCD(int a, int b)
    {
        // ensure that b is always smaller
        if(b > a) return GCD(b, a);

        while (b != 0)
        {
            var temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }
}

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */